pygmt/src/meca.py

```diff
@@ -306,95 +322,20 @@ def update_pointers(data_pointers):
-            (
-                longitude,
-                latitude,
-                depth,
-                plot_longitude,
-                plot_latitude,
-            ) = update_pointers(data_pointers)
-
-            # before constructing the 2D array lets check that each key
-            # of the dict has the same quantity of values to avoid bugs
-            list_length = len(list(spec.values())[0])
-            for value in list(spec.values()):
-                if len(value) != list_length:
-                    raise GMTError(
-                        "Unequal number of focal mechanism "
-                        "parameters supplied in 'spec'."
-                    )
-                # lets also check the inputs for longitude, latitude,
-                # and depth if it is a list or array
-                if (
-                    isinstance(longitude, (list, np.ndarray))
-                    or isinstance(latitude, (list, np.ndarray))
-                    or isinstance(depth, (list, np.ndarray))
-                ):
-                    if (len(longitude) != len(latitude)) or (
-                        len(longitude) != len(depth)
-                    ):
-                        raise GMTError(
-                            "Unequal number of focal mechanism " "locations supplied."
-                        )
-
-            # values are ok, so build the 2D array
-            spec_array = []
-            for index in range(list_length):
-                # Construct the array one row at a time (note that order
-                # matters here, hence the list comprehension!)
-                row = [longitude[index], latitude[index], depth[index]] + [
-                    spec[key][index] for key in foc_params
-                ]
-
-                # Add in plotting options, if given, otherwise add 0s as
-                # required by GMT
-                for arg in plot_longitude, plot_latitude:
-                    if arg is None:
-                        row.append(0)
-                    else:
-                        if "A" not in kwargs:
-                            kwargs["A"] = True
-                        row.append(arg[index])
-                spec_array.append(row)
-            spec = spec_array
-
-        # or assemble the array for the case of pd.DataFrames
-        elif isinstance(dict_type_pointer, np.ndarray):
+            # Convert single int, float data to List[int, float] data
+            _spec = {
+                "longitude": np.atleast_1d(longitude),
+                "latitude": np.atleast_1d(latitude),
+                "depth": np.atleast_1d(depth),
+            }
+            _spec.update({key: np.atleast_1d(val) for key, val in spec.items()})
+            spec = pd.DataFrame.from_dict(_spec)
+
+        assert isinstance(spec, pd.DataFrame)
+        dict_type_pointer = spec.values
+
+        # Assemble the array for the case of pd.DataFrames
+        if isinstance(dict_type_pointer, np.ndarray):
```

Comment:
Just to explain this change (clearer if you look at commit e2230573aa751c61fcb25d997454d3bbcd4e6a45), previously the `spec` could be formatted in 3 different paths depending on the type:

1. `if isinstance(dict_type_pointer, (int, float)):` (i.e. a dictionary of int/float like `{"depth": 0}`
2. `elif isinstance(dict_type_pointer, list):` (i.e. a dictionary of lists like `{"depth": [0, 1, 2]}`
3. `elif isinstance(dict_type_pointer, np.ndarray):` (i.e. a pandas.DataFrame with `np.ndarray` columns)

There was a lot of duplication in these 3 code paths, so I've removed paths 1 and 2 by converting any `spec` provided by the user in dictionary format into `pandas.DataFrame`. So now, everything happens via path 3 only.

Another benefit of handling everything using pandas.DataFrame is that columns can have different data types, so this sets things up for the ability to pass in string labels later in a follow up PR.